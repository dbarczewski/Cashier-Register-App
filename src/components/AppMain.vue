<template>
  <div class="flex flex-grow flex-col justify-between">
    <div class="bg-gray-100 flex-grow flex justify-center items-center">
      <main class="mt-10 bg-white w-1/2 shadow">
        <button
          class="relative w-full h-32 flex justify-center items-center"
          @click="toggleState(0)"
        >
          <span class="absolute left-4 top-4">{{ content.payLabel }}</span>
          <div
            class="text-3xl"
            :class="
              selectedState === 0
                ? 'border-b-8 p-2 border-double border-green-400'
                : ''
            "
          >
            <span>{{ convertNumberToEUR(payAmount) }}</span>
          </div>
        </button>
        <hr />
        <button
          class="relative w-full h-32 flex justify-center items-center"
          @click="toggleState(1)"
        >
          <span class="absolute left-4 top-4">{{ content.givenLabel }}</span>
          <span
            v-if="paymentError"
            class="absolute right-4 top-4 text-red-500"
            >{{ content.error.givenAmountTooLow }}</span
          >
          <div
            class="text-3xl"
            :class="
              selectedState === 1
                ? 'border-b-8 p-2 border-double border-green-400'
                : ''
            "
          >
            <span>{{ convertNumberToEUR(givenAmount) }}</span>
          </div>
        </button>
      </main>
    </div>
    <div
      class="bg-gray-700 flex justify-center py-2 relative space-x-4 shadow-xl"
    >
      <possible-amounts
        @input-number="handlePossibleAmountsNumber"
        :values="calculatePossibleAmounts(payAmount)"
      />
      <numeric-key-pad @input-number="handleKeyPadNumber" />
      <button
        class="
          absolute
          right-4
          bottom-4
          px-6
          py-1
          bg-blue-400
          border-2 border-white
          text-white text-3xl
          shadow-2xl
        "
        @click="triggerPayment"
      >
        {{ content.payButton }}
      </button>
      <div
        class="
          absolute
          top-0
          left-1/2
          -translate-x-1/2
          transform
          -translate-y-full
          flex
          space-x-px
          shadow-xl
        "
      >
        <button
          v-for="x in content.paymentOptions"
          :key="x"
          class="px-10 py-2 shadow-xl"
          :class="
            x === 'Bar'
              ? 'bg-gray-700 text-gray-200'
              : 'bg-gray-300 text-gray-700'
          "
        >
          {{ x }}
        </button>
      </div>
    </div>
    <app-modal
      @close="closeModal"
      v-if="showModal"
      class="
        absolute
        transform
        -translate-x-1/2 -translate-y-1/2
        top-1/2
        left-1/2
      "
    >
      {{ convertNumberToEUR(this.givenAmount - this.payAmount) }} is the change
    </app-modal>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import NumericKeyPad from "./NumericKeyPad.vue";
import PossibleAmounts from "./PossibleAmounts.vue";
import AppModal from "./AppModal.vue";
import Content from "./../../content.json";

import {
  convertNumberToEUR,
  calculatePossibleAmounts,
  addToNumber,
  removeLastDigit,
} from "@/utils/CurrencyConverter";

enum STATE {
  PAY,
  GIVE,
}

export default defineComponent({
  components: { NumericKeyPad, PossibleAmounts, AppModal },
  name: "Main",
  data() {
    return {
      selectedState: STATE.PAY as STATE,
      paymentError: false,
      showModal: false,
      payAmount: "0.00",
      givenAmount: "0.00",
      content: Content,
    };
  },
  mounted() {
    window.addEventListener("keydown", (e) => {
      if (e.key === "Enter") {
        if (this.selectedState === STATE.PAY) {
          this.selectedState = STATE.GIVE;
        } else if (this.selectedState === STATE.GIVE) {
          this.triggerPayment();
        }
      }
    });
  },
  methods: {
    convertNumberToEUR,
    calculatePossibleAmounts,
    addToNumber,
    removeLastDigit,
    closeModal() {
      this.showModal = false;
      this.selectedState = STATE.PAY;
      this.resetAmounts();
    },
    resetAmounts() {
      this.payAmount = "0.00";
      this.givenAmount = "0.00";
    },
    toggleState(state: STATE) {
      this.selectedState = state;
    },
    handleKeyPadNumber(value: number | string) {
      this.paymentError = false;
      let stateToChange =
        this.selectedState === STATE.PAY ? this.payAmount : this.givenAmount;
      if (value === "<") {
        stateToChange = removeLastDigit(stateToChange);
      } else {
        stateToChange = addToNumber(stateToChange, value as string);
      }
      if (this.selectedState === STATE.PAY) {
        this.payAmount = stateToChange;
      } else {
        this.givenAmount = stateToChange;
      }
    },
    handlePossibleAmountsNumber(value: number) {
      console.log("emitted", value);
      this.givenAmount = "" + value;
    },
    triggerPayment() {
      const parsedPayAmount = parseFloat(this.payAmount);
      const parsedGivenAmount = parseFloat(this.givenAmount);
      console.log(this.payAmount);
      console.log(this.givenAmount);

      if (parsedPayAmount > parsedGivenAmount) {
        this.paymentError = true;
      } else {
        this.showModal = true;
      }
    },
  },
});
</script>
