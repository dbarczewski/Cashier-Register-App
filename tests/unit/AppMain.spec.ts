import { shallowMount, mount } from "@vue/test-utils";
import AppMain from "@/components/AppMain.vue";

describe("AppMain", () => {
  it("renders component correctly", () => {
    const wrapper = shallowMount(AppMain);
    expect(wrapper.element).toMatchSnapshot();
  });
  it("checks if eventListener is being removed after unmount", async () => {
    const addEventListenerSpy = jest.spyOn(global, "addEventListener");
    const removeEventListenerSpy = jest.spyOn(global, "removeEventListener");
    const wrapper = mount(AppMain);

    expect(addEventListenerSpy).toBeCalled();
    expect(removeEventListenerSpy).not.toBeCalled();
    wrapper.unmount();
    expect(wrapper.emitted().inputNumber).toBeFalsy();
  });
  it("should switch states or triggersPayment when enter is being pressed", () => {
    const wrapper = mount(AppMain);
    const triggerPaymentSpy = jest.spyOn(wrapper.vm, "triggerPayment");
    let selectedState = wrapper.vm.selectedState;
    expect(selectedState).toBe(0);
    wrapper.vm.handleKeyDownEvent(
      new KeyboardEvent("keydown", {
        key: "Enter",
      })
    );
    expect(triggerPaymentSpy).not.toBeCalled();
    selectedState = wrapper.vm.selectedState;
    expect(selectedState).toBe(1);
    expect(triggerPaymentSpy).not.toBeCalled();
    wrapper.vm.handleKeyDownEvent(
      new KeyboardEvent("keydown", {
        key: "Enter",
      })
    );
    expect(triggerPaymentSpy).toBeCalled();
  });
  it("should handle the keypad numbers correctly for the pay state", () => {
    const wrapper = mount(AppMain);
    wrapper.vm.handleKeyPadNumber(1);
    expect(wrapper.vm.payAmount).toBe("00.01");
    wrapper.vm.handleKeyPadNumber(2);
    expect(wrapper.vm.payAmount).toBe("000.12");
    wrapper.vm.handleKeyPadNumber("delete");
    expect(wrapper.vm.payAmount).toBe("00.01");
    wrapper.vm.handleKeyPadNumber(2);
    wrapper.vm.handleKeyPadNumber(3);
    expect(wrapper.vm.payAmount).toBe("0001.23");
  });
  it("should handle the keypad numbers correctly for the given state", () => {
    const wrapper = mount(AppMain);
    wrapper.vm.toggleState(1);
    wrapper.vm.handleKeyPadNumber(1);
    expect(wrapper.vm.givenAmount).toBe("00.01");
    wrapper.vm.handleKeyPadNumber(2);
    expect(wrapper.vm.givenAmount).toBe("000.12");
    wrapper.vm.handleKeyPadNumber(3);
    expect(wrapper.vm.givenAmount).toBe("0001.23");
  });
  it("should set error if the payment is not sufficient", () => {
    const wrapper = mount(AppMain);
    wrapper.setData({
      givenAmount: "1.00",
      payAmount: "2.00",
    });
    expect(wrapper.vm.paymentError).toBeFalsy();
    wrapper.vm.triggerPayment();
    expect(wrapper.vm.paymentError).toBeTruthy();
  });
  it("should reset the amounts", () => {
    const wrapper = mount(AppMain);
    wrapper.setData({
      givenAmount: "1.00",
      payAmount: "2.00",
    });
    wrapper.vm.resetAmounts();
    expect(wrapper.vm.givenAmount).toBe("0.00");
    expect(wrapper.vm.payAmount).toBe("0.00");
  });
  it("should handle possibleAmountTrigger correctly", () => {
    const wrapper = mount(AppMain);
    wrapper.vm.handlePossibleAmountsNumber(100);
    expect(wrapper.vm.givenAmount).toBe("100");
  });
  it("should close the modal", () => {
    const wrapper = mount(AppMain);
    wrapper.setData({
      showModal: true,
      selectedState: 1,
    });
    expect(wrapper.vm.showModal).toBeTruthy();
    expect(wrapper.vm.selectedState).toBe(1);
    wrapper.vm.closeModal();
    expect(wrapper.vm.showModal).toBeFalsy();
    expect(wrapper.vm.selectedState).toBe(0);
  });
});

/**
 * handleKeyPadNumber(value: number | string) {
      this.paymentError = false;
      let stateToChange =
        this.selectedState === STATE.PAY ? this.payAmount : this.givenAmount;
      if (value === "<") {
        stateToChange = removeLastDigit(stateToChange);
      } else {
        stateToChange = addToNumber(stateToChange, value as string);
      }
      if (this.selectedState === STATE.PAY) {
        this.payAmount = stateToChange;
      } else {
        this.givenAmount = stateToChange;
      }
    },
 */
