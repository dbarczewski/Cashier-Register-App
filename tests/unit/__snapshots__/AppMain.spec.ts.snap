// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppMain renders component correctly 1`] = `
<div
  class="flex flex-grow flex-col justify-between"
>
  <div
    class="bg-gray-100 flex-grow flex justify-center items-center"
  >
    <main
      class="mt-10 bg-white w-1/2 shadow"
    >
      <button
        class="relative w-full h-32 flex justify-center items-center"
      >
        <span
          class="absolute left-4 top-4"
        >
          To pay
        </span>
        <div
          class="text-3xl border-b-8 p-2 border-double border-green-400"
        >
          <span>
            0,00 €
          </span>
        </div>
      </button>
      <hr />
      <button
        class="relative w-full h-32 flex justify-center items-center"
      >
        <span
          class="absolute left-4 top-4"
        >
          Received
        </span>
        <!--v-if-->
        <div
          class="text-3xl"
        >
          <span>
            0,00 €
          </span>
        </div>
      </button>
    </main>
  </div>
  <div
    class="bg-gray-700 flex justify-center py-2 relative space-x-4 shadow-xl"
  >
    <possible-amounts-stub
      values="0.00,1,2,5,10"
    />
    <numeric-key-pad-stub />
    <button
      class="
          absolute
          right-4
          bottom-4
          px-6
          py-1
          bg-blue-400
          border-2 border-white
          text-white text-3xl
          shadow-2xl
        "
    >
      Pay
    </button>
    <div
      class="
          absolute
          top-0
          left-1/2
          -translate-x-1/2
          transform
          -translate-y-full
          flex
          space-x-px
          shadow-xl
        "
    >
      
      <button
        class="px-10 py-2 shadow-xl bg-gray-300 text-gray-700"
      >
        Cash
      </button>
      <button
        class="px-10 py-2 shadow-xl bg-gray-300 text-gray-700"
      >
        EC
      </button>
      <button
        class="px-10 py-2 shadow-xl bg-gray-300 text-gray-700"
      >
        Creditcard
      </button>
      <button
        class="px-10 py-2 shadow-xl bg-gray-300 text-gray-700"
      >
        Coupon
      </button>
      <button
        class="px-10 py-2 shadow-xl bg-gray-300 text-gray-700"
      >
        Others
      </button>
      
    </div>
  </div>
  <!--v-if-->
</div>
`;
